### PR TITLE
raftstore: clean callback when peer is destroyed

### PR DESCRIPTION
**Pull request by BusyJay**
Callback of read index isn't cleared when the peer is destroyed, which will lead to dead lock.

@siddontang @zhangjinpeng1987 @hhkbp2 PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
CI failed

File: tests/raftstore/test_lease_read.rs [hunk 1]
```diff
@@ -380,3 +385,37 @@ fn test_server_lease_unsafe_during_leader_transfers() {
     let mut cluster = new_node_cluster(0, count);
     test_lease_unsafe_during_leader_transfers(&mut cluster);
 }
+
+#[test]
```
Comment (hhkbp2) [inline tests/raftstore/test_lease_read.rs]: Add a brief description of what this test case is for.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -760,12 +776,10 @@ impl Peer {
...
         // Note that only after handle read_states can we identify what requests are
         // actually stale.
-        if let Some(ref ss) = ready.ss {
-            if ss.raft_state != StateRole::Leader {
-                let term = self.term();
-                // all uncommitted reads will be dropped silently in raft.
-                self.pending_reads.clear_uncommitted(&self.tag, term);
-            }
+        if ready.ss.is_some() {
+            let term = self.term();
```
Comment (AndreMouche) [inline src/raftstore/store/peer.rs]: Why not clear for leader before?

**Comment by hhkbp2** [comment]
Rest LGTM

File: tests/raftstore/transport_simulate.rs [hunk 3]
```diff
@@ -294,11 +293,20 @@ impl Filter<RaftMessage> for RegionPacketFilter {
                (self.direction.is_send() && self.store_id == from_store_id ||
                 self.direction.is_recv() && self.store_id == to_store_id) &&
                self.msg_type.as_ref().map_or(true, |t| t == &m.get_message().get_msg_type()) {
-                if self.allow.load(Ordering::Relaxed) > 0 {
-                    self.allow.fetch_sub(1, Ordering::Relaxed);
-                    return true;
-                }
-                return false;
+                return match self.block {
```
Comment (AndreMouche) [inline tests/raftstore/transport_simulate.rs]: Can we add some annotations in this file, it's hard for me to understand?

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -71,8 +79,8 @@ struct ReadIndexQueue {
 
 impl ReadIndexQueue {
     fn clear_uncommitted(&mut self, tag: &str, term: u64) {
-        for read in self.reads.drain(self.ready_cnt..) {
-            for (req, cb) in read.cmds {
+        for mut read in self.reads.drain(self.ready_cnt..) {
+            for (req, cb) in read.cmds.drain(..) {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: When drop is implemented, it can be moved partially.

File: src/raftstore/store/peer.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: When the peer is not leader before, there will be no uncommitted in pending reads. Remove the check just make code more simple.

File: tests/raftstore/transport_simulate.rs (hunk 3 above)
Comment (BusyJay) [inline tests/raftstore/transport_simulate.rs]: Sure, but not related to this pr.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**